readme.md
# Moddable Helper

![Build Status](https://github.com/johnwargo/moddable-helper/actions/workflows/main.yml/badge.svg)

## Introduction

Delivers a command-line interface to simplifying usage of the [Moddable SDK](https://github.com/Moddable-OpenSource/moddable). This repository hosts a node.js module that:

* Delivers a simple set of commands that drive the Moddable SDK, and eliminates some of the repetitive typing required to work with the Moddable SDK command-line tools
* Uses a project configuration file to configure the settings modules and target devices used by the project
* Delivers a mechanism that enables developers to more easily wipe devices using the device's native SDK (without having to remember the command every time)

The module supports the following commands:

* `mddbl init` - Create an empty configuration file in the current folder
* `mddbl config edit` - Edit the project's configuration file
* `mddbl config show` - Display the contents of the configuration file
* `mddbl config sort` - Sort the module and target lists by name
* `mddbl debug` - Toggle the `debug` setting (true to false, and, of course, false to true) in the project's configuration file
* `mddbl deploy [module] [target]` - Deploy the selected module (module or host) to the specified target device
* `mddbl module add <module>` - Adds an empty Module entry to the configuration file's `modules` array, then launches the configuration file in the system's default editor.
* `mddbl module rm <module>` - Removes the specified module from the configuration file's `modules` array
* `mddbl module show <module>` - Shows the configuration settings for the specified module
* `mddbl modules` - Display the list of modules defined in the `modules` section of the configuration file
* `mddbl target add <target>` - Adds an empty Target entry to the configuration file's `targets` array, then launches the configuration file in the system's default editor.
* `mddbl target rm <target>` - Removes the specified target from the configuration file's `targets` array
* `mddbl target show <target>` - Shows the configuration settings for the specified target
* `mddbl targets` - Display the list of targets defined in the `targets` section of the configuration file
* `mddbl wipe <target>` - Wipe the selected target device

> **Note**: Required parameters are in angled brackets (`<` and `>`) and option parameters are in square brackets (`[` and `]`)

Detailed instructions for these commands sections that follow.

## Installation

To install the module, open a terminal window (or command prompt on Windows), and execute the following command:

```shell
npm install -g moddable-helper
```

This installs a `mddbl` command you can use anywhere on the system.

## Issues & Pull Requests

If you have an issue with this module, don't email the author, submit an [issue](https://github.com/johnwargo/moddable-helper/issues) in this repository instead.

I built this project for the community, so if you have an enhancement, fix, or change for the project please submit a [Pull Request](https://github.com/johnwargo/moddable-helper/pulls).

## Background

I've been playing around lately with the Moddable SDK and I think the platform's pretty interesting for a lot of reasons I'll write about later on [my blog](https://johnwargo.com). They deliver several very capable IoT devices with built-in displays at a great price. The real power comes from their SDK which delivers a robust and solid JavaScript API for inexpensive microcontroller devices

One of the things I noticed when I started working with the Moddable SDK was that I found myself typing the same commands over and over again as I coded projects for the devices. Now, that's not unexpected for command-line based tooling, but when building and deploying for Moddable devices, I kept typing the exact same command-line options over and over again (`-d m -p <device>`). I decided to make this helper module to reduce my typing and simplify my work.

Let me show you how this works using a real-world example.

The folks at Moddable published a book called [IoT Development for ESP32 and ESP8266 with JavaScript](https://github.com/Moddable-OpenSource/iot-product-dev-book) that contains a lot of sample code demonstrating how to use the different APIs in their SDK. Moddable developers often break a project down into multiple parts, a host (the core native application running on the device plus a little JavaScript code to bootstrap the project) plus additional JavaScript modules that run within the host (a dramatic oversimplification, I know, but I'm not here to teach you Moddable development). Developers typically keep their host and module files in different folders within a project folder.

> **Note**: What I show here applies to any Moddable SDK project

Using the book's Chapter 1 [Hello World](https://github.com/Moddable-OpenSource/iot-product-dev-book/tree/master/ch1-gettingstarted) project as an example, to deploy the host to a Moddable Two device, you will open a terminal window, navigate to the project folder, then execute the following commands:

```shell
cd host
mcconfig -d -m -p esp32/moddable_two
```

Next, to deploy the host, execute the following commands:

```shell
cd ../helloworld
mcrun -d -m -p esp32/moddable_two
```

As you can see, I'm switching folders from time to time depending on whether I'm deploying an update to the project's host or an update to the `helloworld` module. I'm also executing two different Moddable SDK commands, each with the exact same command-line parameters. There has to be a better way.

Yes, I know I could open two different terminal windows and execute host or module commands in their own terminal - I'm for simplification here, not opening even more windows.

With this module, you create a configuration file called `mddbl.json` (the module can create it for your automatically using the `mddbl init` command) in the project's root folder, then define all of the options for your project there:

```json
{
  "debug": true,
  "modules": [
    {
      "name": "host",
      "description": "The project's host module",
      "isHost": true,
      "debugFlag": true,
      "makeFlag": true,
      "folderPath": "host"
    },
    {
      "name": "hw",
      "description": "Text-only version of the project",
      "isHost": false,
      "debugFlag": true,
      "makeFlag": true,
      "folderPath": "helloworld"
    },
    {
      "name": "hwg",
      "description": "Hello World Graphical version",
      "isHost": false,
      "debugFlag": true,
      "makeFlag": true,
      "folderPath": "helloworld-gui"
    }
  ],
  "targets": [
    {
      "name": "mdbl2",
      "description": "Moddable Two",
      "platform": "esp32/moddable_two",
      "formatFlag": false,
      "formatStr": "",
      "rotationFlag": false,
      "rotationValue": 0,
      "wipeCommand": "python %IDF_PATH%\\components\\esptool_py\\esptool\\esptool.py erase_flash"
    },
    {
      "name": "m5fire",
      "description": "M5Stack Fire device",
      "platform": "esp32/m5stack_fire",
      "formatFlag": true,
      "formatStr": "gray16",
      "rotationFlag": false,
      "rotationValue": 0,
      "wipeCommand": "python %IDF_PATH%\\components\\esptool_py\\esptool\\esptool.py erase_flash"
    }
  ]
}
```

With that in place, you can open a terminal window, navigate to your project folder, then deploy the project's host using:

```shell  
mddbl deploy host mdbl2
```

The mddbl module switches to the host module's folder (`host` as specified in the module's `folderPath` property), executes the command to deploy the host (`mcconfig -d -m -p esp32/moddable_two`), then switches back to the starting folder.

Next, to deploy one of the project's modules, `helloworld` for example, simply use the following command:

```shell
mddbl deploy hw mdbl2
```

Moddable Helper switches to the module's folder (`helloworld` as specified in the module's `folderPath` property), executes the command to deploy the host (`mcrun -d -m -p esp32/moddable_two`), then switches back to the starting folder.

> **Note**: The mddbl module uses the module's `isHost` property to determine whether to execute `mcconfig` or `mcrun` to deploy the module.

## Usage

### Configuration

To use this module, you must first create the module's configuration file in the Moddable project's root folder. The file must be called `mddbl.json` and the default configuration file contains the following options:

```json
{
  "debug": false,
  "modules": [],
  "targets": [],
}
```

#### Creating the Configuration File

The module can create it for you automatically, simply open a terminal window, navigate to the project folder, and execute the following command:

```shell
mddbl init
```

Moddable Helper will create the default configuration file shown above.

#### Configuration Options

The following sections describe the configuration options for the Moddable Helper module.

##### Debug

If you're having trouble with the Moddable Helper, you can enable debug mode through the configuration file. In debug mode, the CLI outputs more information to the console as it runs.

To enable debug mode, set the configuration file's `debug` option to `true` as shown in the following example:

```json
{
  "debug": true,
  "modules": [],
  "targets": [],
}
```

You can also toggle the `debug` setting using the following command:

```shell
mddbl debug
```

##### Modules

Use the `modules` section of the configuration file to configure an array of `module` objects representing each of the JavaScript modules in your Moddable project. The `module` object has the following configuration:

```json
{
  "name": "",
  "description": "",
  "isHost": false,
  "debugFlag": true,
  "makeFlag": true,
  "folderPath": "",
}
```

* `name` - The unique identifier for the module. This is the value you'll use in `mddbl` commands to refer to the module.
* `description` - A description of the module, not used for anything except to remind you about this module's configuration.
* `isHost` - Boolean value indicating whether the module is a Host or Module (controls whether Moddable Helper executes `mcconfig` or `mcrun` to deploy the module).
* `debugFlag` - Enables/disables the `-d` parameter passed to `mcconfig` and `mcrun`; refer to the [Moddable documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/tools/tools.md#arguments) for details about this parameter.
* `makeFlag` - Enables/disables the `-m` parameter passed to `mcconfig` and `mcrun`; refer to the [Moddable documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/tools/tools.md#arguments) for details about this parameter.
* `folderPath` - The name of the subfolder hosting the module.

The example configuration file shown above defines three modules:

```json
"modules": [
  {
    "name": "host",
    "description": "The project's host module",
    "isHost": true,
    "debugFlag": true,
    "makeFlag": true,
    "folderPath": "host"
  },
  {
    "name": "hw",
    "description": "Text-only version of the project",
    "isHost": false,
    "debugFlag": true,
    "makeFlag": true,
    "folderPath": "helloworld"
  },
  {
    "name": "hwg",
    "description": "Hello World Graphical version",
    "isHost": false,
    "debugFlag": true,
    "makeFlag": true,
    "folderPath": "helloworld-gui"
  }
],
```

The `host` module refers to the JavaScript Host module code in the project's `host` folder. The `hw` and `hwg` modules refer to JavaScript Modules in the `helloworld` and `helloworld-gui` folders.

##### Targets

Use the `targets` section of the configuration file to configure an array of `target` objects representing each of the Moddable-compatible hardware devices used by your project. The `target` object has the following configuration:

```json
{
  "name": "",
  "description": "",
  "platform": "",
  "wipeCommand": "",
}
```

* `name` - The unique identifier for the device. This is the value you'll use in `mddbl` commands to refer to the device; the Moddable SDK uses a full description to refer to a device (`esp32/moddable_two` for example) and this just gives you a shortcut for the full name.
* `description` - A description of the device, not used for anything except to remind you about this device.
* `platform` - The full Moddable SDK platform identifier for the device (`esp32/moddable_two` for example).
* `formatFlag` - Enables/disables the `-f` parameter passed to `mcconfig` and `mcrun`; refer to the [Moddable documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/tools/tools.md#arguments) for details about this parameter.
* `formatStr` - With `formatFlag` enabled, specifies the format string passed to the Moddable SDK. Available options are (from the Moddable SDK Docs) `gray16`, `gray256`, `rgb332`, `rgb565be` or `rgb565le`. The Moddable SDK defaults to `rgb565le` if you leave this value empty.
* `rotationFlag` - Enables/disables the `-r` parameter passed to `mcconfig` and `mcrun`; refer to the [Moddable documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/tools/tools.md#arguments) for details about this parameter.
* `rotationValue` - With `rotationFlag` enabled, specifies the screen rotation value. Supported values are `0`, `90`, `180` or `270`. The Moddable SDK defaults to 0, but there's no way to leave this value blank, so you must ensure a valid value is set for this property.
* `wipeCommand` - The file system command to wipe the device. The command used is specific to the hardware platform, not the Moddable SDK.

> **Note**: On Microsoft Windows, you must double-up file system delimiters; `"python %IDF_PATH%\components\esptool_py\esptool\esptool.py erase_flash"` becomes `"python %IDF_PATH%\\components\\esptool_py\\esptool\\esptool.py erase_flash"`. JavaScript uses the backslash (`\`) when escaping other characters in a string, so to include the backslash in a command string, you must escape it with a backslash first. On Linux or macOS, the command string is "python $IDF_PATH/components/esptool_py/esptool/esptool.py erase_flash" (no special escaping needed).

The example configuration file shown above defines two Targets devices:

```json
"targets": [
  {
    "name": "mdbl2",
    "description": "Moddable Two",
    "platform": "esp32/moddable_two",
    "formatFlag": false,
    "formatStr": "",
    "rotationFlag": false,
    "rotationValue": 0,
    "wipeCommand": "python %IDF_PATH%\\components\\esptool_py\\esptool\\esptool.py erase_flash"
  },
  {
    "name": "m5fire",
    "description": "M5Stack Fire device",
    "platform": "esp32/m5stack_fire",
    "formatFlag": true,
    "formatStr": "gray16",
    "rotationFlag": false,
    "rotationValue": 0,
    "wipeCommand": "python %IDF_PATH%\\components\\esptool_py\\esptool\\esptool.py erase_flash"
  }
]
```

> **Note**: The targets use the same wipe command because the devices are both based on the ESP32 hardware platform.

#### Viewing the Configuration File

To view the contents of the configuration file in the console, execute the following command:

```shell
mddbl config show
```

#### Editing the Configuration File

To launch the module's configuration file using your system's default `.json` file editor, execute the following command:

```shell
mdbbl config edit
```

#### Sorting Module and Target Lists

For users with a bit of obsessive compulsive disorder (OCD), you can sort the Module and Target lists using the following command:

```shell
mdbbl config sort
```

You know, just because.

### Deploy

To deploy a Host or Module to a connected device, execute the following command in the terminal window pointing to the Moddable project folder:

```shell
mddbl deploy <module> [target]
```

> **Note**: The `target` parameter is optional. If you omit it, then the Moddable SDK will load the module in the emulator for the operating system running the command (`win` for Windows, `mac` for macOS, etc.).


changelog.md
# Changelog

## 0.0.11 - May 20, 2021

* Added a `finally` clause to the `doDeploy` function, ensuring it switches back to the starting folder

## 0.0.10 - May 11, 2021

* Reverted the `deployModule` changes from 0.0.9
* Actually fixed the issue with `deploy`; it now properly processes the deployment when you don't provide a target

## 0.0.9 - May 11, 2021

* Fixed logic error in `deploy` command, didn't fail correctly if a target was not passed to the command
* Refactored (simplified) the `checkDirectory` function

## 0.0.8 - May 10, 2021

* Added required parameter `module` to the `module add` command
* Added required parameter `target` to the `target add` command

## 0.0.7 - May 9, 2021

Mother's Day Edition

* Added `module` commands `add`, `rm`, `show`
* Added `target` commands `add`, `rm`, `show`
* Refactored most function names and grouped related functions

## 0.0.6 - May 6, 2021

* Fixed issue with config edit on macOS

## 0.0.5 - May 5, 2021

* Readme file changes

## 0.0.4 - May 4, 2021

* Added interactive deployment mode

## 0.0.3 - May 4, 2021

* Added the `formatFlag` and `formatStr` configuration options
* Added the `rotationFlag` and `rotationValue` configuration options
* Added the `modules` and `targets` commands
* Removed the `list` commands
* Removed the `platformFlag` configuration option (that didn't last long, did it?)

## 0.0.2 - May 4, 2021

* Added the `platformFlag` configuration option
* Updated the `.npmignore` file to keep some sample files from the install

## 0.0.1 - May 3, 2021

Initial release


mddbl.js
#!/usr/bin/env node
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var config_1 = require("./config");
var chalk = require('chalk');
var fs = require('fs');
var logger = require('cli-logger');
var os = require('os');
var path = require('path');
var program = require('commander');
var Select = require('enquirer').Select;
var cp = require("child_process");
var packageDotJSON = require('./package.json');
var APP_NAME = '\nModdable Helper (mddbl)';
var APP_AUTHOR = 'by John M. Wargo (https://johnwargo.com)';
var CONFIG_FILE_NAME = 'mddbl.json';
var CHECK_CONFIG_STRING = "not defined, please check the module configuration (" + CONFIG_FILE_NAME + ")";
var WORKING_PATH = process.cwd();
var appConfig;
var configFilePath;
var log = logger();
function checkDirectory(filePath) {
    log.debug("Locating " + filePath);
    if (fs.existsSync(filePath)) {
        try {
            var stats = fs.statSync(filePath);
            if (stats)
                return stats.isDirectory;
        }
        catch (err) {
            log.error("checkDirectory error: " + err);
        }
    }
    return false;
}
function showObjectProperties(itemName, type) {
    log.debug("showObjectValues(" + itemName + ", " + type + ")");
    configRead();
    var theArray = type === 'Module' ? appConfig.modules : appConfig.targets;
    var obj = theArray.find(function (item) { return item.name === itemName; });
    if (!obj) {
        log.error(type + " '" + itemName + "' " + CHECK_CONFIG_STRING);
        return;
    }
    log.info("Configuration for the '" + obj.name + "' " + type + ":");
    console.dir(obj);
}
function listArrayNames(listStr, theList) {
    log.debug('listArray()');
    if (theList.length > 0) {
        log.info("\nConfigured " + listStr + ":");
        for (var item in theList) {
            var outputStr = "- " + theList[item].name;
            if (theList[item].description) {
                outputStr += " - " + theList[item].description;
            }
            log.info(outputStr);
        }
    }
    else {
        log.info("\nNo " + listStr + " configured");
        process.exit(1);
    }
}
function deleteArrayItem(typeStr, theArray, compareStr) {
    var idx = -1;
    for (var i = 0; i < theArray.length; i++) {
        if (idx < 0) {
            log.debug(i + ": " + theArray[i].name);
            if (theArray[i].name == compareStr) {
                log.debug("Found match at index " + idx);
                idx = i;
            }
        }
    }
    if (idx > -1) {
        theArray.splice(idx, 1);
    }
    else {
        log.error(typeStr + " '" + compareStr + "' " + CHECK_CONFIG_STRING);
        process.exit(1);
    }
    return theArray;
}
function configEdit() {
    log.info('Editing module configuration');
    console.log(configFilePath);
    var cmdStr = (os.type().indexOf('Win') === 0)
        ? "start " + CONFIG_FILE_NAME
        : "open -e ./" + CONFIG_FILE_NAME;
    cp.exec(cmdStr, function (error, stdout, stderr) {
        if (error) {
            log.error('Unable to edit configuration');
            log.error(error);
            process.exit(1);
        }
        if (stdout) {
            log.info(stdout);
        }
        if (stderr) {
            log.error(stderr);
        }
    });
}
function configInit() {
    log.info('Initializing project folder...');
    appConfig = Object.assign({}, config_1.defaultConfig);
    if (configWrite()) {
        log.info("Successfully created configuration file (" + CONFIG_FILE_NAME + ")");
    }
    else {
        process.exit(1);
    }
}
function configRead() {
    log.info('Reading configuration file');
    if (fs.existsSync(configFilePath)) {
        try {
            var rawData = fs.readFileSync(configFilePath);
            appConfig = JSON.parse(rawData);
        }
        catch (err) {
            log.error("readConfig error: " + err);
            process.exit(1);
        }
        var logLevel = appConfig.debug ? log.DEBUG : log.INFO;
        log.level(logLevel);
        log.debug('\nProgram Information (debug)');
        log.debug(APP_AUTHOR);
        log.debug("Version: " + packageDotJSON.version);
        log.debug('Command Options:', program.opts());
        log.debug("Working directory: " + WORKING_PATH);
        log.debug("Configuration file: " + configFilePath + "\n");
    }
    else {
        log.info("\nConfiguration file not found (" + configFilePath + ")");
        log.info("Execute " + chalk.yellow('`mdbbl config init`') + " to create one here");
        process.exit(1);
    }
}
function configWrite() {
    log.info("Writing configuration to " + configFilePath);
    var data = JSON.stringify(appConfig, null, 2);
    try {
        fs.writeFileSync(configFilePath, data);
        log.debug('Configuration file successfully written to disk');
    }
    catch (err) {
        log.error('Unable to write to configuration file');
        log.error(err);
        return false;
    }
    return true;
}
function configShow() {
    log.debug('showConfig()');
    configRead();
    log.info('\nModule configuration:');
    console.dir(appConfig);
}
function configSort() {
    function compare(a, b) {
        return a.name > b.name ? 1 : -1;
    }
    configRead();
    if (appConfig) {
        appConfig.modules.sort(compare);
        appConfig.targets.sort(compare);
        if (!configWrite()) {
            process.exit(1);
        }
    }
}
function debugToggle() {
    log.debug('toggleDebug()');
    configRead();
    if (appConfig) {
        log.debug("Toggling Debug configuration parameter to " + !appConfig.debug);
        appConfig.debug = !appConfig.debug;
        if (!configWrite()) {
            process.exit(1);
        }
    }
}
function doDeploy(rootCmd, mod, target) {
    if (target) {
        log.debug("doDeploy(" + rootCmd + ", " + mod.name + ", " + target.name + ")");
    }
    else {
        log.debug("doDeploy(" + rootCmd + ", " + mod.name + ")");
    }
    var folder = mod.folderPath;
    if (!checkDirectory(folder)) {
        log.error("Specified module folder (" + folder + ") does not exist");
        process.exit(1);
    }
    var cmd = rootCmd + ' ';
    if (mod.debugFlag)
        cmd += '-d ';
    if (mod.makeFlag)
        cmd += '-m ';
    if (target) {
        cmd += "-p " + target.platform + " ";
        if (target.formatFlag) {
            cmd += "-f ";
            if (target.formatStr.length > 0)
                cmd += target.formatStr + " ";
        }
        if (target.rotationFlag) {
            cmd += "-r ";
            if (target.rotationValue > -1)
                cmd += target.rotationValue + " ";
        }
    }
    log.debug("Command: " + cmd);
    log.info(chalk.yellow("Changing to the '" + folder + "' directory"));
    process.chdir(folder);
    log.debug("Current directory: " + process.cwd());
    try {
        log.info(chalk.yellow('Executing:') + " " + cmd);
        cp.execSync(cmd, { stdio: 'inherit' });
    }
    catch (e) {
        log.error(chalk.red(e));
        process.exit(1);
    }
    finally {
        log.info(chalk.yellow("Changing back to the '" + WORKING_PATH + "' directory"));
        process.chdir(WORKING_PATH);
    }
}
function deployModule(modName, targetName) {
    log.debug("deployModule(" + modName + ", " + targetName + ")");
    if (!appConfig)
        configRead();
    var mod = appConfig.modules.find(function (item) { return item.name === modName; });
    if (!mod) {
        log.error("ERROR: Module '" + modName + "' " + CHECK_CONFIG_STRING);
        process.exit(1);
    }
    if (!mod.folderPath) {
        log.error("ERROR: Module path '" + mod.folderPath + "' " + CHECK_CONFIG_STRING);
        process.exit(1);
    }
    var target;
    if (targetName) {
        target = appConfig.targets.find(function (item) { return item.name === targetName; });
        if (!target) {
            log.error("ERROR: Target '" + targetName + "' " + CHECK_CONFIG_STRING);
            process.exit(1);
        }
        if (!target.platform) {
            log.error("ERROR: Target platform '" + target.platform + "' " + CHECK_CONFIG_STRING);
            process.exit(1);
        }
        if (target.rotationFlag && target.rotationValue &&
            !(target.rotationValue == 0 || target.rotationValue == 90 || target.rotationValue == 180 || target.rotationValue == 270)) {
            log.error("ERROR: Invalid Target rotation value (" + target.rotationValue + ")");
            process.exit(1);
        }
    }
    console.log("Deploying " + modName + " to " + targetName);
    if (mod.isHost) {
        doDeploy('mcconfig', mod, target);
    }
    else {
        doDeploy('mcrun', mod, target);
    }
}
function deployInteractive() {
    return __awaiter(this, void 0, void 0, function () {
        var modPrompt, targetPrompt, modName, targetName;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    log.debug('Deploying in interactive mode');
                    configRead();
                    if (appConfig.modules.length < 1) {
                        log.error("Module list " + CHECK_CONFIG_STRING);
                        process.exit(1);
                    }
                    if (appConfig.targets.length < 1) {
                        log.error("Module list " + CHECK_CONFIG_STRING);
                        process.exit(1);
                    }
                    modPrompt = new Select({
                        name: 'modName',
                        message: 'Module Selection',
                        choices: appConfig.modules
                    });
                    targetPrompt = new Select({
                        name: 'targetName',
                        message: 'Target Selection',
                        choices: appConfig.targets
                    });
                    modName = '';
                    targetName = '';
                    return [4, modPrompt.run()
                            .then(function (result) {
                            modName = result.toString();
                        })
                            .catch(function (err) {
                            log.error(err);
                            process.exit(1);
                        })];
                case 1:
                    _a.sent();
                    return [4, targetPrompt.run()
                            .then(function (result) {
                            targetName = result.toString();
                        })
                            .catch(function (err) {
                            log.error(err);
                            process.exit(1);
                        })];
                case 2:
                    _a.sent();
                    if (modName.length > 0 && targetName.length > 0) {
                        deployModule(modName, targetName);
                    }
                    return [2];
            }
        });
    });
}
function moduleAdd(modName) {
    return __awaiter(this, void 0, void 0, function () {
        var newMod;
        return __generator(this, function (_a) {
            log.debug("moduleAdd(" + modName + ")");
            configRead();
            newMod = Object.assign({}, config_1.emptyModule);
            newMod.name = modName;
            appConfig.modules.push(newMod);
            configWrite();
            configEdit();
            return [2];
        });
    });
}
function moduleRemove(modName) {
    log.debug("moduleRemove(" + modName + ")");
    configRead();
    appConfig.modules = deleteArrayItem('Module', appConfig.modules, modName);
    configWrite();
}
function moduleShow(modName) {
    log.debug("moduleShow(" + modName + ")");
    showObjectProperties(modName, 'Module');
}
function modulesList() {
    log.debug('modulesList()');
    configRead();
    listArrayNames('Modules', appConfig.modules);
}
function targetAdd(targetName) {
    log.debug('targetAdd()');
    configRead();
    var newTarget = Object.assign({}, config_1.emptyTarget);
    newTarget.name = targetName;
    appConfig.targets.push(newTarget);
    configWrite();
    configEdit();
}
function targetRemove(targetName) {
    log.debug("targetRemove(" + targetName + ")");
    configRead();
    appConfig.targets = deleteArrayItem('Target', appConfig.targets, targetName);
    configWrite();
}
function targetShow(targetName) {
    log.debug("targetShow(" + targetName + ")");
    showObjectProperties(targetName, 'Target');
}
function targetsList() {
    configRead();
    listArrayNames('Targets', appConfig.targets);
}
function wipeDevice(targetName) {
    log.debug("wipeDevice(" + targetName + ")");
    configRead();
    var target = appConfig.targets.find(function (item) { return item.name === targetName; });
    if (!target) {
        log.error("Target '" + targetName + "' " + CHECK_CONFIG_STRING);
        process.exit(1);
    }
    if (!target.wipeCommand) {
        log.error("Target wipe command '" + target.wipeCommand + "' " + CHECK_CONFIG_STRING);
        process.exit(1);
    }
    try {
        console.log("Wiping " + targetName);
        cp.execSync(target.wipeCommand, { stdio: 'inherit' });
    }
    catch (e) {
        log.error(chalk.red(e));
        process.exit(1);
    }
}
console.log(APP_NAME);
configFilePath = path.join(WORKING_PATH, CONFIG_FILE_NAME);
program.version(packageDotJSON.version);
program.option('--debug', 'Output extra information during operation');
var configCmd = program.command('config')
    .description("Work with the module's configuration");
configCmd
    .command('edit')
    .description('Edit the module\'s configuration file')
    .action(configEdit);
configCmd
    .command('show')
    .description('Print the modules config to the console')
    .action(configShow);
configCmd
    .command('sort')
    .description('Sorts the config modules and targets arrays')
    .action(configSort);
program
    .command('debug')
    .description('Toggle the debug configuration setting')
    .action(debugToggle);
program
    .command('deploy [module] [target]')
    .description('Deploy Module to specified Target device')
    .action(function (module, target) {
    if (module) {
        deployModule(module, target);
    }
    else {
        deployInteractive();
    }
});
program
    .command('init')
    .description('Initialize the current folder (create module config file)')
    .action(configInit);
var moduleCmd = program.command('module')
    .description('Work with the modules configuration');
moduleCmd
    .command('add <module>')
    .description('Add an empty module to the configuration file')
    .action(function (module) {
    moduleAdd(module);
});
moduleCmd
    .command('rm <module>')
    .description('Remove a module from the configuration file')
    .action(function (module) {
    moduleRemove(module);
});
moduleCmd
    .command('show <module>')
    .description('Show a module configuration')
    .action(function (module) {
    moduleShow(module);
});
program
    .command('modules')
    .description('List all configured Modules')
    .action(modulesList);
var targetCmd = program.command('target')
    .description('Work with the targets configuration');
targetCmd
    .command('add <target>')
    .description('Add an empty target to the configuration file')
    .action(function (target) {
    targetAdd(target);
});
targetCmd
    .command('rm <target>')
    .description('Remove a target from the configuration file')
    .action(function (target) {
    targetRemove(target);
});
targetCmd
    .command('show <target>')
    .description('Show a target configuration')
    .action(function (target) {
    targetShow(target);
});
program
    .command('targets')
    .description('List all configured Targets')
    .action(targetsList);
program
    .command('wipe <target>')
    .description('Wipes the <target> device')
    .action(function (target) {
    wipeDevice(target);
});
program.parse();
